### PR TITLE
Addressing ECAL local reconstruction on GPU issues

### PR DESCRIPTION
**Pull request by thomreis**
#### PR description:

This PR addresses some issues that have been left open during the initial PR #31719 . They are tracked in issue #32480 .
The fixed issues:
- Update the ECAL ESProducts and move them to a more correct place
- Migrate GPU code to use common constants and functions
- Some cleanup in code comments

#### PR validation:

Passes matrix WFs 10824.512 and 11634.512.
No changes in the RecHits output.

### CODE REVIEW

**Comment by thomreis** [comment]
@amassiro @vkhristenko FYI

**Comment by perrotta** [comment]
enable gpu

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by thomreis** [comment]
Failures include also many non-ECAL related histograms. Perhaps an issue with WF 11634.911? PR #33126 seems to have similar test results.

**Comment by perrotta** [comment]
> Failures include also many non-ECAL related histograms. Perhaps an issue with WF 11634.911? PR #33126 seems to have similar test results.

The isssues with wf 11634.911 are known, and are not expected to depend on this PR, see #32963

**Comment by jpata** [comment]
@thomreis I was going through the changes, and I see that a number of files have been moved with `git mv` and thus the changes show up cleanly, but a number of files (e.g. EcalTimeBiasCorrectionsGPU.h) show up as a removed and readded file. I just wanted to ask - would it be possible to use `git mv` on those files too, to better preserve the history? Or perhaps you did, and github (or me) is just getting confused. If not, and you don't see this as an undue burden, I would like to ask to rewrite the history to use `git mv` as much as possible, so we can better see what the changes in the files are.

Thanks!

**Comment by thomreis** [comment]
@jpata I did not move any files with `git mv`. But I committed each header that I have moved separately to preserve this history. Also `EcalTimeBiasCorrectionsGPU.h` has it own commit (https://github.com/cms-sw/cmssw/pull/33116/commits/2ffdda7bb696b28e8182ce5172dfdc910ad5082e), where it shows as moved. However, since some headers were substantially changed after moving in the three commits named "Replace cudaMallow and cudaMemcpyAsync..." git gets confused and does not recognise them anymore in the overall files changed tab.
I am not sure if `git mv` would perform better in this situation.

**Comment by yuanchao** [comment]
+1
* matrix tests and unit tests all passed, as well as the GPU addon tests.
* static build warnings mainly come from the deprecated es<>.get() (none part of this PR)

File: CondFormats/EcalObjects/interface/EcalRecHitParametersGPU.h [hunk 1]
```diff
@@ -13,9 +15,10 @@
 class EcalRecHitParametersGPU {
 public:
   struct Product {
-    ~Product();
-    int *ChannelStatusToBeExcluded, *expanded_v_DB_reco_flags;
-    uint32_t *expanded_Sizes_v_DB_reco_flags, *expanded_flagbit_v_DB_reco_flags;
+    edm::propagate_const_array<cms::cuda::device::unique_ptr<int[]>> ChannelStatusToBeExcluded;
```
Comment (jpata) [inline CondFormats/EcalObjects/interface/EcalRecHitParametersGPU.h:18]:
```suggestion
    edm::propagate_const_array<cms::cuda::device::unique_ptr<int[]>> channelStatusToBeExcluded;
``

File: CondFormats/EcalObjects/interface/EcalTimeBiasCorrectionsGPU.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#endif  // __CUDACC__
+
+class EcalTimeBiasCorrectionsGPU {
+public:
+  struct Product {
+    edm::propagate_const_array<cms::cuda::device::unique_ptr<float[]>> EBTimeCorrAmplitudeBins;
+    edm::propagate_const_array<cms::cuda::device::unique_ptr<float[]>> EBTimeCorrShiftBins;
+    edm::propagate_const_array<cms::cuda::device::unique_ptr<float[]>> EETimeCorrAmplitudeBins;
+    edm::propagate_const_array<cms::cuda::device::unique_ptr<float[]>> EETimeCorrShiftBins;
+    int EBTimeCorrAmplitudeBinsSize, EETimeCorrAmplitudeBinsSize;
```
Comment (jpata) [inline CondFormats/EcalObjects/interface/EcalTimeBiasCorrectionsGPU.h:20]: EB, EE -> eb, ee to stick with https://cms-sw.github.io/cms_coding_rules.html 2.9 (sorry)

File: CondFormats/EcalObjects/interface/EcalTimeBiasCorrectionsGPU.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+  //
+  static std::string name() { return std::string{"ecalTimeBiasCorrectionsGPU"}; }
+#endif  // __CUDACC__
+
+private:
+  std::vector<float, cms::cuda::HostAllocator<float>> EBTimeCorrAmplitudeBins_;
+  std::vector<float, cms::cuda::HostAllocator<float>> EBTimeCorrShiftBins_;
+  std::vector<float, cms::cuda::HostAllocator<float>> EETimeCorrAmplitudeBins_;
+  std::vector<float, cms::cuda::HostAllocator<float>> EETimeCorrShiftBins_;
```
Comment (jpata) [inline CondFormats/EcalObjects/interface/EcalTimeBiasCorrectionsGPU.h:42]: EB, EE -> eb, ee

File: CondFormats/EcalObjects/src/EcalRecHitParametersGPU.cc [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  //     https://github.com/cms-sw/cmssw/blob/266e21cfc9eb409b093e4cf064f4c0a24c6ac293/RecoLocalCalo/EcalRecProducers/plugins/EcalRecHitWorkerSimple.cc
+
+  // Traslate string representation of flagsMapDBReco into enum values
+  const edm::ParameterSet& p = ps.getParameter<edm::ParameterSet>("flagsMapDBReco");
+  std::vector<std::string> recoflagbitsStrings = p.getParameterNames();
+
+  for (unsigned int i = 0; i != recoflagbitsStrings.size(); ++i) {
+    EcalRecHit::Flags recoflagbit = (EcalRecHit::Flags)StringToEnumValue<EcalRecHit::Flags>(recoflagbitsStrings[i]);
+    std::vector<std::string> dbstatus_s = p.getParameter<std::vector<std::string>>(recoflagbitsStrings[i]);
+    //     std::vector<uint32_t> dbstatuses;
```
Comment (jpata) [inline CondFormats/EcalObjects/src/EcalRecHitParametersGPU.cc:25]: still needed?

**Review by jpata — COMMENTED** [review]
Thanks for this PR - I went over the changes and just some minor style things.
Did you check that there are no changes to the GPU outputs?

**Comment by thomreis** [comment]
The tests were including the GPU tests and no differences were reported in those WFs.

**Comment by fwyzard** [comment]
> @thomreis I was going through the changes, and I see that a number of files have been moved with `git mv` and thus the changes show up cleanly, but a number of files (e.g. EcalTimeBiasCorrectionsGPU.h) show up as a removed and readded file. I just wanted to ask - would it be possible to use `git mv` on those files too, to better preserve the history? Or perhaps you did, and github (or me) is just getting confused. If not, and you don't see this as an undue burden, I would like to ask to rewrite the history to use `git mv` as much as possible, so we can better see what the changes in the files are.

I don't think that `git mv A B` does anything different than `mv A B; git rm A; git add B`.

Compare:
```bash
fwyzard@Fool:~/test/git_mv$ git mv A B
fwyzard@Fool:~/test/git_mv$ git status
On branch master
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        renamed:    A -> B

fwyzard@Fool:~/test/git_mv$ git status --no-renames
On branch master
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        deleted:    A
        new file:   B
```
with:
```bash
fwyzard@Fool:~/test/git_mv$ mv A B; git rm A; git add B
rm 'A'
fwyzard@Fool:~/test/git_mv$ git status
On branch master
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        renamed:    A -> B

fwyzard@Fool:~/test/git_mv$ git status --no-renames
On branch master
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        deleted:    A
        new file:   B
```

**Comment by fwyzard** [comment]
please test

**Comment by jpata** [comment]
Thx for the updates. As far as I see, this is the pending question before the reco sig. I can also run a comparison myself, I was just wondering if it was already done. Thanks!

> Did you check that there are no changes to the GPU outputs?

**Comment by thomreis** [comment]
The additional GPU tests seem to consist of WFs 10824.502 and 10824.512 and they passed. However, it looks like the two WFs were not included in the comparisons (https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_3_X_2021-03-16-2300+e8dd18/41754/). Is that the normal behaviour?

**Comment by jpata** [comment]
There is no automatic comparison of GPU workflows at the moment, it needs to be implemented. I can run a check wrt. the baseline (or if you already did, let me know).

**Comment by thomreis** [comment]
I am running a comparison with 10824.512 atm.

**Comment by thomreis** [comment]
Comparing WF 10824.512 ECAL sorted RecHits from CMSSW_11_3_0_pre3 with the ones from CMSSW_11_3_0_pre3 + this PR on 100 events, I get the same number of RecHits with the same number of mean and std dev. values for all variables. For EB and EE. So this looks like there are no changes in the output, as expected.

**Comment by jpata** [comment]
+reconstruction
- cleanup and ESProduct migration of the ECAL GPU code, partly addresses #32480
- technical, no reco changes expected or observed (a manual check was run in https://github.com/cms-sw/cmssw/pull/33116#issuecomment-802042750)

**Comment by thomreis** [comment]
@cms-sw/db-l2 , @cms-sw/alca-l2 do you have questions or comments?

**Comment by jpata** [comment]
@cms-sw/db-l2 @cms-sw/alca-l2

**Comment by ggovi** [comment]
+1

**Comment by silviodonato** [comment]
kind reminder @cms-sw/alca-l2

**Comment by yuanchao** [comment]
+1
* PR updated to comply the coding rules
* matrix tests and unit tests all passed, as well as the GPU addon tests.
* static build warnings mainly come from the deprecated es<>.get() (none part of this PR)

**Comment by qliphy** [comment]
+1

**Pull request merged**